File: src/code/python/02_linked_list/04_swap_pairs/swap_pairs.py
import unittest

class ListNode:
    def __init__(self, val: int = 0, next=None) -> None:
        self.val = val
        self.next = next
        
class LinkedList:
    """
    两两交换链表中的节点 https://leetcode.cn/problems/swap-nodes-in-pairs/
    """
    
    def __init__(self) -> None:
        self.head = ListNode()
        self.size = 0
        
    def add_at_tail(self, val:int):
        current = self.head
        
        while current.next != None:
            current = current.next
        
        current.next = ListNode(val=val)
        self.size += 1

    def swap_pairs(self) ->ListNode:
        current = self.head
        
        while current.next != None and current.next.next != None:
            n = current.next
            nn = current.next.next
            
            n.next = nn.next
            nn.next = n
            current.next = nn
            
            current = current.next.next
        
        return self.head.next

class TestSwapPairs(unittest.TestCase):
    def test_swap_pairs(self):
        list = LinkedList()
        list.add_at_tail(1)
        list.add_at_tail(2)
        list.add_at_tail(3)
        list.add_at_tail(4)
        
        res = list.swap_pairs()
        
        while res != None:
            print(res.val)
            res = res.next
            
if __name__ == "__main__":
    unittest.main()

File: src/code/php/02_linked_list/04_swap_pairs/swap_pairs.php
<?php

class ListNode
{
    public function __construct(
        public int $val = 0,
        public ListNode|NULL $next = NULL
    ) {
    }
}

/**
 * 两两交换链表中的节点 https://leetcode.cn/problems/swap-nodes-in-pairs/
 */
class LinkedList
{
    public ListNode $head;
    public int $size;

    public function __construct()
    {
        $this->head = new ListNode(0);
        $this->size = 0;
    }

    public function addAtTail(int $val)
    {
        $current = $this->head;
        while ($current->next != NULL) {
            $current = $current->next;
        }

        $current->next = new ListNode($val);
        $this->size++;
    }

    public function swapPairs(): ListNode
    {
        $current = $this->head;
        while ($current->next != NULL && $current->next->next != NULL) {
            $n = $current->next;
            $nn = $current->next->next;

            $n->next = $nn->next;
            $nn->next = $n;
            $current->next = $nn;

            $current = $current->next->next;
        }

        return $this->head->next;
    }
}

$list = new LinkedList();
$list->addAtTail(1);
$list->addAtTail(2);
$list->addAtTail(3);
$list->addAtTail(4);
$list->addAtTail(5);
$res = $list->swapPairs();

while ($res != NULL) {
    print($res->val . "\n");
    $res = $res->next;
}
